drivers: stepper: refactor ISR safe event mechanism

jilaypandya opened the pull request:
Refactor ISR safe event handling mechanism from step_dir common to a dedicated module .

jilaypandya committed:
drivers: stepper: drop ISR safe event mechanism from step_dir_common

drop ISR safe event mechanism handling from step_dir_common

Signed-off-by: Jilay Pandya <jilay.pandya@outlook.com>

dipakgmx commented:
I am not sure about the context of the ISR callbacks, so had 2 questions:
1. Why was this check for callback made, if the callback was called in ISR
2. It seems, right now, we are locking the data in spinlock and calling the callback, i mean something like this:
```
K_SPINLOCK(lock) {
stepper_handle_timing_signal -> 
  position_mode_task ->  
    update_remaining_steps -> 
      stepper_trigger_callback -> 
        data->callback
}
```
This, am not sure is the right approach. Guess, the callbacks need to be called from some work, rather than cascading the calls.

jilaypandya committed:
drivers: stepper: introduce stepper event handler

Event handling mechanism was hardcoded in step-dir drivers.
This commit refactors the event handling in a dedicated module.

Signed-off-by: Jilay Pandya <jilay.pandya@outlook.com>

jilaypandya committed:
drivers: stepper: adi: tmc50xx use stepper post event

use stepper_post_event instead of execute callback

Signed-off-by: Jilay Pandya <jilay.pandya@outlook.com>

jilaypandya commented on drivers/stepper/event_handler/stepper_event_handler.c:
```diff
@@ -0,0 +1,82 @@
+	return 0;
+}
+
+SYS_INIT(stepper_event_handler_init, APPLICATION, CONFIG_APPLICATION_INIT_PRIORITY);
```
do you mean the init_prio should be CONFIG_STEPPER_INIT_PRIORITY?

dipakgmx commented on drivers/stepper/event_handler/iterables.ld:
```diff
@@ -0,0 +1,3 @@
+# iterables.ld
+#include <zephyr/linker/iterable_sections.h>
+ITERABLE_SECTION_RAM(stepper_event_handler, Z_LINK_ITERABLE_SUBALIGN)
```
Any drawback of having the `struct stepper_event_handler`, as a data member of the stepper driver? Then the members could be initialised during the driver init and at the same time use the common function from `stepper_event_handler.c`.

jilaypandya replied:
Yeah but then each and every driver would have to do this explicitly. And in this way it happens in the background.

Also with CONFIG_STEPPER_EVENT_HANDLER=n, you could switch off this functionality and the callback would just simply be called without going via the `event_handler`

faxe1008 commented on drivers/stepper/event_handler/stepper_event_handler.c:
```diff
@@ -0,0 +1,82 @@
+	return 0;
+}
+
+SYS_INIT(stepper_event_handler_init, APPLICATION, CONFIG_STEPPER_INIT_PRIORITY);
```
This feels off imo, the driver should be ready for use after `POST_KERNEL` init level has run. I get the idea behind it, that drivers do not have to explicitly call an init function but the reduction in complexity would be worth it.

faxe1008 reviewed (commented):
👍 for moving out the event handling to another file, but I feel like having this inside of an iterable section is a bit too much complexity for few gains. The only real advantage that I see is that the user does not have to call the `stepper_event_handler_init` manually.

jilaypandya commented:
> 👍 for moving out the event handling to another file, but I feel like having this inside of an iterable section is a bit too much complexity for few gains. The only real advantage that I see is that the user does not have to call the `stepper_event_handler_init` manually.

👍 I'll close the PR :)
